provider/overrides/runner.rb

```diff
@@ -0,0 +1,179 @@
+    #     properties:
+    #       someProperty: !ruby/object:Provider::MyProvider::PropertyOverride
+    #         description: 'foobar' # replaces description
+    #       anotherProperty.someNestedProperty:
+    #         !ruby/object:Provider::MyProvider::PropertyOverride
+    #         description: 'baz'
+    #       anotherProperty[].someNestedPropertyInAnArray:
+    #         !ruby/object:Provider::MyProvider::PropertyOverride
+    #         description: 'baz'
+    #   ...
+    class Runner
+      def initialize(api, overrides,
+                     res_override_class = Provider::Overrides::ResourceOverride,
+                     prop_override_class = Provider::Overrides::PropertyOverride)
+        @api = api
+        @overrides = overrides
+        @res_override_class = res_override_class
+        @prop_override_class = prop_override_class
+      end
+
+      def build
+        validator = Provider::Overrides::Validator.new(@api, @overrides)
+        validator.run
+        build_product(@api, @overrides)
+      end
+
+      private
+
+      # Given a old Api::Product, and Provider::Overrides::ResourceOverrides,
+      # returns a new Api::Product with overrides applied
+      def build_product(old_prod, all_overrides)
+        prod = Api::Product.new
+        old_prod.instance_variables
+                .reject { |o| o == :@objects }.each do |var_name|
+          if (all_overrides['product'] || {})[var_name]
+            prod.instance_variable_set(var_name, all_overrides['product'][var_name])
+          else
+            prod.instance_variable_set(var_name, old_prod.instance_variable_get(var_name))
+          end
+        end
+        prod.instance_variable_set('@objects',
+                                   old_prod.objects
+                                           .map { |o| build_resource(o, all_overrides[o.name]) })
+        prod
+      end
+
+      # Given a Api::Resource and Provider::Override::ResourceOverride,
+      # return a new Api::Resource with overrides applied.
+      def build_resource(old_resource, res_override)
+        res_override = @res_override_class.new if res_override.nil? || res_override.empty?
+        res_override.validate
+        res_override.apply old_resource
+
+        res = Api::Resource.new
+        set_additional_values(res, res_override)
+
+        variables = (old_resource.instance_variables + res_override.instance_variables).uniq
+        variables.reject { |o| %i[@properties @parameters].include?(o) }
+                 .each do |var_name|
+          if !res_override[var_name].nil?
+            res.instance_variable_set(var_name, res_override[var_name])
+          else
+            res.instance_variable_set(var_name, old_resource.instance_variable_get(var_name))
+          end
+        end
+
+        # Using instance_variable_get('properties') to make sure we get `exclude: true` properties
+        ['@properties', '@parameters'].each do |val|
```

Comment:
I think previously we'd have to define `disk.encryption` as an override before we defined `disk.encryption.raw_key` otherwise it would try and insert a property into an object that didn't exist yet. If I'm following you correctly this means we will only have to define `disk.encryption.raw_key` and possibly remove some superfluous overrides?
Also I think I gave you a brief overview of how the disk fields are abusing overrides in order to work, but I imagine there will need to be some refactoring before they could use this system because they are super order dependent right now.

Reply:
Yes, that should work.